ui/src/main/webapp/src/app/reports/technologies/tech-report.service.ts

```diff
@@ -119,6 +131,58 @@ export class TechReportService extends GraphService
         return result;
     }
 
+    protected getDataFromFilteredEndpoint<T>(url: string, filter: any): Observable<T[]> {
+        const jsonFilter = this.serializeFilter(filter);
+
+        return this._http.post(url, jsonFilter, this.JSON_OPTIONS)
+            .map(response => {
+                const json = response.json();
+                const entities = this._graphJsonToModelService.fromJSONarray(json);
+                return entities;
+            });
+    }
+
+    getHibernateEntityModel(execID: number, filter?: ReportFilter): Observable<HibernateEntityModel[]> {
+        const url = TechReportService.HIBERNATE_ENTITY_URL.replace('{execID}', execID.toString());
+
+        const entitiesObservable = this.getDataFromFilteredEndpoint<HibernateEntityModel>(url, filter);
+
+        return Observables.resolveValuesArray(entitiesObservable, ['javaClass']).flatMap(entitiesArray => {
+            if (entitiesArray.length === 0) {
+                return Observable.of([]);
+            }
+
+            return Observable.forkJoin(entitiesArray.map(entity => Observables.resolveObjectProperties(entity.resolved.javaClass, ['decompiledSource'])))
+                .map(resolvedJavaClasses => {
+                    const updatedEntitiesArray = [ ... entitiesArray ];
+
+                    return updatedEntitiesArray.map((entity, index)  => {
+                        entity.resolved.javaClass = resolvedJavaClasses[index];
+
+                        return entity;
```

Comment:
TBH, for me it's not been so simple to read and understand this way of handling the "resolved" fields.
I see how it works but i don't have a clear perception about why it's better than the "subscribe" approach.

Other than this, avoiding the usage of DTO is good in this case where there's a 1:1 mapping between data retrieved and data displayed in the UI but in the case of EJB report, not having a DTO would have implied having a lot of if-else to understand where the value to be displayed was stored.
For example, if EJB is defined in a Java class, then you have the vertexId if the source file in decompiledSource but if EJB is defined in XML file, then you'll find the vertexId in the deploymentDescriptor. Also finding the interface is another case of this type because it depends if the EJB is local or remote.

I know that this code works but trying to find a common way of doing things, i think this is a good point to be discussed.

Reply:
I agree it is not ideal. This is not so different from`subscribe` approach, but I just wanted to reduce amount of boilerplate in code, so I created `Observables.resolveValuesArray`. The problem (in both cases) is, when you need to resolve some nested entity. Then it gets really ugly. I admit, in this case, it is probably little bit uglier :( 

But I think this is general problem with our approach to data model. We are trying to have as universal API as possible, but at the end, I think it costs us more pain then gain. Ideal solution would be to have something like GraphQL, where you would specify what data you need, and backend would return it (completely resolved). But that would be quite problematic to create and it would probably take huge amount of effort.

I know originally we had in mind idea that we don't want to write specific endpoint for each report, but rather use some universal API which would allow us to get data from graph. But we are not following this idea anymore anyway. Right now we write endpoint for (almost) each report (`HardcodedIPResource`, `EJBResource`, `HibernateResource`, ...). It almost feels like we ended with approach which gets cons from both worlds (universal and purpose-made API), but very little pros :(

And for DTOs: 
EJB report seems to be special case, where you managed to get all data into one universal DTO class. But in other reports, it is not generally true. For example Remote Services report has several different types of entities.

Reply:
To be honest, I don't see how GraphQL would solve anything. The problem isn't the data itself, but in creating nice DTOs for each report. With GraphQL, you would still have to fully define the query and still have to define the DTO.

In general, I feel like we've spent many more hours talking about this than would have been needed to just write a handful of DTOs and be done with it. It doesn't even matter that much where the DTOs are (client vs, server).